### PR TITLE
New resource: aws_ec2_managed_prefix_list

### LINKED ISSUE #13986: Support New Resource and Data Source for EC2 Managed Prefix Lists
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

> Virtual Private Cloud (VPC) customers can now create and manage their own Prefix Lists to simplify VPC configurations.

##### New

* `aws_ec2_managed_prefix_list`

##### Affected

* `aws_route_table`
* `aws_route`

### Potential Terraform Configuration

```hcl
resource "aws_ec2_managed_prefix_list" "example" {
  address_family = "IPv4"
  max_entries    = 1
  
  entry {
    cidr        = "0.0.0.0/0"
    description = "Everywhere"
  }
}
```

### References

[Announcement](https://aws.amazon.com/about-aws/whats-new/2020/06/amazon-virtual-private-cloud-customers-use-prefix-lists-simplify-configuration-security-groups-route-tables/).
[User guide](https://docs.aws.amazon.com/vpc/latest/userguide/managed-prefix-lists.html).

Requires [AWS SDK v1.32.12](https://github.com/aws/aws-sdk-go/releases/tag/v1.32.12):
* https://github.com/terraform-providers/terraform-provider-aws/pull/13803

### PR DESCRIPTION
**Pull request by roberth-k**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #13986

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
FEATURES:
- **New Resource:** `aws_ec2_managed_prefix_list`

ENHANCEMENTS:
- resource/aws_security_group: Update documentation with managed prefix lists in mind
- resource/aws_security_group_rule: Update documentation with managed prefix lists in mind
```

Update to route tables expected with https://github.com/terraform-providers/terraform-provider-aws/pull/14013.

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS=-run=TestAccAwsEc2ManagedPrefixList
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsEc2ManagedPrefixList -timeout 120m
--- PASS: TestAccAwsEc2ManagedPrefixList_disappears (20.64s)
--- PASS: TestAccAwsEc2ManagedPrefixList_exceedLimit (28.56s)
--- PASS: TestAccAwsEc2ManagedPrefixList_name (38.95s)
--- PASS: TestAccAwsEc2ManagedPrefixList_basic (41.02s)
--- PASS: TestAccAwsEc2ManagedPrefixList_tags (78.83s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       80.429s
```

### CODE REVIEW

**Comment by ewbankkit** [comment]
@roberth-k Thanks for the great contribution 👏.
A few high-level comments:
* As the EC2 APIs are separate, I suggest having a separate `aws_managed_prefix_list` data source and calling out in the documentation that this is for _customer-managed prefix lists_ and the existing `aws_prefix_list` data source is for _AWS-managed prefix lists_
* Split the new data source into a separate PR (dependent on the resource PR as you'll need the corresponding resource for acceptance tests) - Helps with speed of review
* The fixes to the current `aws_prefix_list` data source (thanks) can then be in another separate PR (once again, aids review)
* Given that the [same EC2 API](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_ModifyManagedPrefixList.html) updates both the PL name and the entries in the PL, are there use cases for having a separate `aws_managed_prefix_list_entry` resources?

**Comment by roberth-k** [comment]
Thanks @ewbankkit.

> * As the EC2 APIs are separate, I suggest having a separate `aws_managed_prefix_list` data source and calling out in the documentation that this is for _customer-managed prefix lists_ and the existing `aws_prefix_list` data source is for _AWS-managed prefix lists_
> * Split the new data source into a separate PR (dependent on the resource PR as you'll need the corresponding resource for acceptance tests) - Helps with speed of review
> * The fixes to the current `aws_prefix_list` data source (thanks) can then be in another separate PR (once again, aids review)

The DescribeManagedPrefixLists API is a superset of DescribePrefixLists, even returning AWS's own prefix lists by default. A separate `aws_managed_prefix_list` resource would have to artificially insert a filter by `owner-id` in order to produce only customer-managed prefix lists. It seemed appropriate to present an unambiguous data source, though I did at one point consider aliasing the two names.

Existing uses of `data.aws_prefix_list` would be affected only where no filters had been provided to the data source, in which case the current behavior is to produce an arbitrary AWS prefix list, but after this PR would be to raise an error.

However, it is true that the two API-s are physically separate, so I'm happy to split it out if that's the preferred practice. In any case, I'll move these changes to a separate PR.

> * Given that the [same EC2 API](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_ModifyManagedPrefixList.html) updates both the PL name and the entries in the PL, are there use cases for having a separate `aws_managed_prefix_list_entry` resources?

`aws_managed_prefix_list_entry` follows the example of security groups and routes, allowing the prefix list to be populated from other Terraform modules.

It's true that there's no separate API for it, and implementing the `_entry` resource did necessitate trading off support for the `version` attribute on the main resource (as adding or removing entries would affect it). However: the versioning aspect of Managed Prefix Lists behaves similar to e.g. RevisionId in Lambda, in that it's mainly used as a race detector, and the "restore" functionality still increments it. Ultimately the practical value of `_entry` seemed worth it.

**Comment by roberth-k** [comment]
The data source changes have been split into https://github.com/terraform-providers/terraform-provider-aws/pull/14109 and https://github.com/terraform-providers/terraform-provider-aws/pull/14110.

**Comment by dturnbu** [comment]
Do we have any idea when this will be merged and made available for use?

**Comment by roberth-k** [comment]
Per https://github.com/terraform-providers/terraform-provider-aws/issues/12690#issuecomment-655091416 it may be a while before this is picked up.

In the meanwhile, I recommend using the `aws_cloudformation_stack` resource to wrap an [AWS::EC2::PrefixList](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-prefixlist.html) with [DeletionPolicy](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-deletionpolicy.html) `Retain` for eventual import into Terraform.

**Comment by roberth-k** [comment]
The resources have been renamed to `aws_prefix_list` and `aws_prefix_list_entry`, and the data source changes updated accordingly.

> The new Managed Prefix List API is a superset of the regular Prefix List API. Customer and AWS prefix lists share the same ID format and namespace, and CloudFormation's AWS::EC2::PrefixList resource does not differentiate between managed and non-managed. I've therefore elected to drop most references to prefix lists being managed. The resource was originally named aws_managed_prefix_list, and this remains reflected in the branch name.

**Comment by teamterraform** [comment]
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

**Comment by t0rr3sp3dr0** [comment]
Any updates on this?

**Review by rafaelleonardocruz — APPROVED** [review]

**Comment by bflad** [comment]
Hi @roberth-k 👋  Thank you for starting this work for supporting EC2 Managed Prefix Lists. I'll be leading up the review feedback and actively monitoring this pull request until its merged or handled in some fashion.

It looks like there are some immediately outstanding items, noted by the merge conflicts and [automated message above](https://github.com/hashicorp/terraform-provider-aws/pull/14068#issuecomment-675728633). Are you able to address these?

There is also some other larger feedback that we will want to work through here, but to properly align everyone's efforts and expectations, are you available to handle requested changes before the end of the month? If yes, I will be happy to followup with those items and context. If not (and no pressure), the maintainer team can take your commits and apply changes on top.

Please let me know and thanks again for this contribution!

**Comment by roberth-k** [comment]
Hi @bflad; thank you for getting in touch!

This PR has been behind for a while, as it was a struggle keeping up with master. With firm knowledge about timing, I'm happy to work with the maintainers to get it ready, and will sort out the immediate issues in the next few days.

Is the data source implementation (https://github.com/hashicorp/terraform-provider-aws/pull/14110, https://github.com/hashicorp/terraform-provider-aws/pull/14109) slated for merge as well?

**Comment by bflad** [comment]
> Is the data source implementation (#14110, #14109) slated for merge as well?

I will separately reach out in those PRs -- the functionality will be reviewed and handled in some fashion as part of including this new functionality as well.

File: aws/resource_aws_ec2_managed_prefix_list_test.go
Comment (bflad) [inline aws/resource_aws_ec2_managed_prefix_list_test.go:282]: We should randomize the naming of these resources -- please see the [Contributing Guide](https://github.com/hashicorp/terraform-provider-aws/blob/master/docs/contributing/running-and-writing-acceptance-tests.md#randomized-naming) for information about this (or searching for `rName` usage elsewhere).

File: aws/resource_aws_ec2_managed_prefix_list_test.go
Comment (bflad) [inline aws/resource_aws_ec2_managed_prefix_list_test.go:392]: Similarly here, regarding naming randomization.

File: aws/resource_aws_ec2_managed_prefix_list_test.go
Comment (bflad) [inline aws/resource_aws_ec2_managed_prefix_list_test.go:406]: Similarly here, regarding using `ImportStateVerify` instead.

File: aws/resource_aws_ec2_managed_prefix_list_test.go
Comment (bflad) [inline aws/resource_aws_ec2_managed_prefix_list_test.go:470]: Similarly here, regarding using `ImportStateVerify` instead.

File: website/docs/r/ec2_managed_prefix_list.html.markdown
Comment (bflad) [inline website/docs/r/ec2_managed_prefix_list.html.markdown:19]: Won't be needed:

```suggestion
```

**Review by bflad — CHANGES_REQUESTED** [review]
Hi @roberth-k 👋  Thank you for updating this pull request. I have provided some initial feedback here, with the biggest item to remove the `aws_ec2_managed_prefix_list_entry` resource for now so we can focus on getting the `aws_ec2_managed_prefix_list` resource merged in. Please reach out with any questions or let us know if you will not have time to make these updates. 👍

**Comment by bflad** [comment]
Please note that I've updated the milestone for this pull request to version 3.22.0, which we expect to release on Thursday as our last release before the new year so its not lingering over the long break. If we do not see the updates prior to Wednesday, we will add commits on top of the existing ones to ensure this is tested and merged in time. 👍

**Comment by roberth-k** [comment]
@bflad I'll get the changes in now.

**Comment by bflad** [comment]
Thank you for your work on this, @roberth-k, pulling in your commits now. 👍

**Pull request merged**
